docs/authentication.md

```diff
@@ -2,208 +2,356 @@
-        uint32 time;
-        SignatureRSV rsv;
-    }
+On the client side, the TypeScript code running inside a browser needs to make
+sure that the session token for making authenticated calls is valid. If not, the
+browser requests a wallet to sign a log-in message and fetches a fresh session
+token.
+
+```typescript
+import {SiweMessage} from 'siwe';
+
+async function getSecretMessage(): Promise<Message> {
+  const messageBox = await hre.ethers.getContractAt('MessageBox', '0x5FbDB2315678afecb367f032d93F642f64180aa3');
+
+  if (token == '') { // Stored in browser session.
+    const domain = await messageBox.domain();
+    const siweMsg = new SiweMessage({
+      domain,
+      address: addr, // User's selected account address.
+      uri: `http://${domain}`,
+      version: "1",
+      chainId: 0x5afe, // Sapphire Testnet
+    }).toMessage();
+    const sig = Signature.from((await window.ethereum.getSigner(addr)).signMessage(siweMsg));
+    token = await messageBox.login(siweMsg, sig);
+  }
+
+  return messageBox.getSecretMessage(bearer);
+}
+```
 
-    modifier authenticated(SignIn calldata auth)
-    {
-        // Must be signed within 24 hours ago.
-        require( auth.time > (block.timestamp - (60*60*24)) );
+:::info Example: Starter project
 
-        // Validate EIP-712 sign-in authentication.
-        bytes32 authdataDigest = keccak256(abi.encodePacked(
-            "\x19\x01",
-            DOMAIN_SEPARATOR,
-            keccak256(abi.encode(
-                SIGNIN_TYPEHASH,
-                auth.user,
-                auth.time
-            ))
-        ));
+To see a running example of the TypeScript SIWE code including the Hardhat
+tests, Node.js and the browser, check out the official Oasis [demo-starter]
+project.
 
-        address recovered_address = ecrecover(
-            authdataDigest, uint8(auth.rsv.v), auth.rsv.r, auth.rsv.s);
+:::
 
-        require( auth.user == recovered_address, "Invalid Sign-In" );
+:::tip Sapphire TypeScript wrapper?
 
-        _;
+While the [Sapphire TypeScript wrapper][sp-npm] offers a convenient end-to-end
+encryption for the contract calls out of the box, it is not mandatory for SIWE.
+
+:::
+
+[demo-starter]: https://github.com/oasisprotocol/demo-starter/tree/matevz/sapphire-paratime-2.0
+[EIP-4361]: https://eips.ethereum.org/EIPS/eip-4361
+[single sign-on]: https://en.wikipedia.org/wiki/Single_sign-on
+[sp-npm]: https://www.npmjs.com/package/@oasisprotocol/sapphire-paratime
+
+### via signed queries
+
+The [EIP-712] proposal defines a method to show data to the user in a structured
+fashion so they can verify it and sign it. In the frontend apps signing a view
+call would require user interaction each time—sometimes even multiple times per
+page—which results in bad UX. Backend services on the other hand can have direct
+access to an Ethereum wallet without needing user interaction. This is possible
+because such services never access random websites, but execute only trusted
+code.
```

Comment:
```suggestion
The [EIP-712] proposal defines a method to show data to the user in a structured
fashion, so they can verify it and sign it. On the frontend, apps signing a view
call would require user interaction each time—sometimes even multiple times per
page—which is bad UX that frustrates users. Backend services on the other hand can have direct
access to an Ethereum wallet without needing user interaction. This is possible
because these services do not access unspecified websites, and only execute trusted
code.
```